### DISCUSSION TITLE
Lost the ability to install basic.rp6502

### DISCUSSION THREAD

**Discussion by marklinebaugh** [picocomputer/community#121] [Q&A] [answered]
First, I must say this is an awesome project... well done!

A few months ago, I installed basic.rp6502 and upon reboot the BASIC interpreter would boot instead of the OS Monitor. Something seems to have changed. Now, when I "INSTALL" basic, the computer becomes unresponsive. A reboot is required to get back to the OS Monitor prompt. Once I get back to the OS prompt (via REBOOT switch), I am able to execute the "set boot basic" command. After which, the basic ROM will be listed when the HELP command is executed. But, when I REBOOT to start the ROM, I get the error message "?Missing RP6502 ROM header". This behavior also occurs on another video game project I'm working on. 

These are the steps I'm taking... initial state is "no installed ROMs". 1) copy basic.rp6502 to USB drive, 2) install the drive in the picocomputer, 3) INSTALL basic.rp6502, 4) the computer seems to hang, 5) after several minutes with no response (no prompt), I activate the reboot switch , 6) normal reboot occurs, when I execute the HELP command the OS reports "1 installed ROM:" and on the next line "BASIC."  7) I execute the "SET BOOT BASIC" command without error, 8) I reboot via command line or the REBOOT switch and I get the error messaged noted above.  Anyone else seeing something similar? Any suggestions? 

Note: If I "LOAD" the file, there are no issues with either the BASIC interpreter or my game project.

**Comment by rumbledethumps** [comment]
First guess is the littlefs file system may be corrupt or you have a bad/worn flash block, but "set boot basic" also writes to this file system and seems to be working.

Try swapping Picos and/or flash_nuke.uf2:
https://www.raspberrypi.com/documentation/microcontrollers/pico-series.html#resetting-flash-memory

**Reply by marklinebaugh** [reply]
Thanks! Here's what I've learned after applying your recommendations.  

First, I switched to a new USB drive, with no change in results. 

So, I swapped PICOs and nuked the flash on both. I reloaded 0.8 release on VGA with no apparent issues. When I reloaded 0.8 on the RIA, my PC made the "tasked completed" sequence of tones, but the PICO's LED remained off and the following (obviously partial) message was displayed (I presume "Cor" is just part of the work "Corrupted"). 

	Picocomputer 6502
	RIA Version 0.8
	/home/rumble/rp6502/src/littlefs/lfs.c:1346:error: Cor

In summary, the issue is occurring on both PICOs. When I flash either PICO standalone (not installed on the Picomputer board) the results are the same as when they're plugged into the board. And, in case it matters, I get the message above when my PC is connected to either the RIA or the VGA PICO.

Seems particularly odd that the firmware load is somehow corrupted irrespective of the PICO I use and that it occurs when the firmware is loaded with the PICO installed on the Picocomputer board or standalone. Quite the head scratcher...  at least for me :^) 

Thanks for your help!

**Comment by rumbledethumps** [comment]
It looks like the 0.8 binary has something wrong with littlefs.
https://github.com/orgs/picocomputer/discussions/116

Nothing changed from the previous version except TinyUSB which isn't directly connected to littlefs.

**Reply by marklinebaugh** [reply]
Here's a bit more data.... Just tried to revert to 0.7... boot up was normal, including display of RIA and VGA revisions. I was able to INSTALL BASIC, set BOOT to BASIC and reboot to BASIC. However, the the "VGA Version" changed from "0.7" to "Unknown" after installing BASIC. Otherwise, everything seemed to be working. 

But, when I tried to INSTALL my video game code HELLO.RP6502, with BOOT set to HELLO, everything looked normal until I rebooted. After reboot, the normal OS Monitor display is shown with both versions at 0.7 (unlike the BASIC test case), however immediately following the version info the following is displayed, followed by a command prompt.

	?watchdog timeout

FYI - The RP6502 game file I'm installing is fairly large 76.3 kB, while the BASIC file is only 11.9 kB.

Happy to send a zipped RP6502 file containing the game code.

**Reply by marklinebaugh** [reply]
Ahhh... sorry I missed that thread. 

The point made toward the end of the post regarding issues with flashing 0.8 to a new or recently nuked PICO sounded intriguing. As noted, I had already reverted to 0.7, so a did the experiment of loading 0.8 while 0.7 was still resident instead of nuking 0.7 first... and it worked... except I accidentally left the USB hub with keyboard and USB drive disconnected. When I reconnected the USB hub, my code continued to run, but when I rebooted it failed with the same symptoms. When I issued ctl/alt/del, I got the OS prompt. When rebooted from the command line, I got the same symptoms. When I RESET from the command line, the code successfully restarts. If i disconnect the USB Hub, I can successfully reboot to my code (via switch). 

A couple other data points, removing the hub and directly connecting the keyboard didn't help. And, when connecting the keyboard a message is sometimes displayed, not clear on what is causing the variation. If I connected the keyboard while my code was running, the message displayed (if a message was displayed at all) was ''*** PANIC", at the bottom of console screen. If I attached the keyboard (no hub) after ctl/alt/del and resulting command prompt, I get just "*** PAN" on the line after the command prompt.

The RIA USB connection seems to be a key ingredient to the behaviors I'm seeing. Hopefully this is helpful.

**Reply by marklinebaugh** [reply]
UPDATE: I just discovered that my PHI2 setting was at 4000 (4MHz). I had been doing some macro performance eval a few weeks back and failed to reset the PHI2 back to 8000. After this change, I was able to INSTALL and BOOT my game on both the 0.7 and 0.8 revisions with no "watchdog timer" errors. I was able to do this with my USB hub connected and with USB drive and keyboard connected to the hub. Perhaps the INSTALL/BOOT watchdog timing issue is related to the 6502's performance??

**Comment by rumbledethumps** [comment, marked as answer]
I was able to reproduce this. Simply bumping littlefs to the latest version seems to fix it. Released fixed binary as v0.9.
https://github.com/picocomputer/rp6502/releases

**Reply by marklinebaugh** [reply]
Looks good! Thanks!
